common/txmgr/abandoned_tracker.go

```diff
@@ -0,0 +1,194 @@
+	tracker.txes = append(tracker.txes, AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{
+		tx:        tx,
+		fatalTime: time.Now().Add(TTL),
+	})
+}
+
+// markFatal sets a transaction's state to fatal_error
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) markFatal(
+	ctx context.Context,
+	atx AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+	tracker.lggr.Infow(fmt.Sprintf("tx %v marked as fatal for exceeding ttl", atx.tx.ID))
+
+	atx.tx.Error.SetValid(AbandonedErrorMsg)
+
+	err := (*tracker.txStore).UpdateTxFatalError(ctx, atx.tx)
+	if err != nil {
+		tracker.lggr.Errorw(fmt.Sprintf("failed to mark tx %v as fatal", atx.tx.ID))
+		// TODO: Handle error
+	}
+}
+
+// getAbandonedAddresses retrieves fromAddress’s in evm.txes that are not present in the Confirmer's enabledAddresses list
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) getAbandonedAddresses(enabledAddrs []ADDR) ([]ADDR, error) {
+	fromAddresses, err := (*tracker.ks).EnabledAddressesForChain(tracker.chainID)
+	if err != nil {
+		return nil, err
+	}
+
+	var abandoned []ADDR
+	for _, addr := range fromAddresses {
+		if !slices.Contains(enabledAddrs, addr) {
+			abandoned = append(abandoned, addr)
+		}
+	}
+
+	return abandoned, nil
+}
+
+// HandleAbandonedTransactions is called by the Confirmer to track and handle all abandoned transactions
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) HandleAbandonedTransactions(ctx context.Context, enabledAddrs []ADDR) {
+	// Find abandoned addresses
+	abandonedAddrs, err := tracker.getAbandonedAddresses(enabledAddrs)
+	if err != nil {
+		// TODO handle error
+	}
+
+	// Get abandoned txes from addresses and insert into the tracker
+	for _, addr := range abandonedAddrs {
+		seq, err := (*tracker.client).SequenceAt(ctx, addr, nil)
+		if err != nil {
+			// TODO handle error
+		}
+
+		tx, err := (*tracker.txStore).FindTxWithSequence(ctx, addr, seq)
+		if err != nil {
+			// TODO handle error
+		}
+
+		tracker.insertAbandonedTx(tx)
+	}
+
+	// Check states of all current abandoned transactions and update tracking
+	tracker.handleTransactionStates(ctx)
+}
+
+// handleTransactionStates handles all abandoned transactions based on their current state.
+// Transactions with finalized states are no longer tracked, while transactions which
+// exceed their ttl are marked as fatal.
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) handleTransactionStates(ctx context.Context) {
+	temp := make([]AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], 0)
+
+	for _, atx := range tracker.txes {
+		switch atx.tx.State {
+		case TxConfirmed, TxConfirmedMissingReceipt, TxFatalError:
+			// Stop tracking tx when finalized state is obtained
+			continue
+		case TxInProgress:
+			if time.Now().After(atx.fatalTime) {
+				tracker.markFatal(ctx, atx)
+				continue
+			}
+			temp = append(temp, atx)
```

Comment:
You might need a TODO here to confirm the status of the tx on chain in case it was confirmed so we can update the state in our DB. There's probably some logic from the Confirmer we can use here.